src/lambda_function/raw_sync/app.py

```diff
@@ -0,0 +1,518 @@
+        or central_directory_offset + central_directory_size > range_end
+    ):
+        logger.warning(
+            "The entire central directory is not contained in "
+            f"ContentRange {object_response['ContentRange']}."
+        )
+        appropriate_range_size = total_size - central_directory_offset
+        logger.warning(
+            f"Calling this function recursively with `range_size` = {appropriate_range_size}"
+        )
+        return list_files_in_archive(
+            s3_client=s3_client,
+            bucket=bucket,
+            key=key,
+            range_size=appropriate_range_size,
+        )
+
+    # Compile a list of file names which satisfy the same conditions used by dispatch Lambda
+    with zipfile.ZipFile(BytesIO(tail), "r") as zip_file:
+        for zip_info in zip_file.infolist():
+            if (
+                "/" not in zip_info.filename  # necessary for pilot data only
+                and "Manifest" not in zip_info.filename
+                and zip_info.file_size > 0
+            ):
+                file_object = {
+                    "filename": zip_info.filename,
+                    "file_size": zip_info.file_size,
+                }
+                file_list.append(file_object)
+    return file_list
+
+
+def publish_to_sns(
+    bucket: str, key: str, path: str, file_size: int, sns_arn: str
+) -> None:
+    """
+    Publishes file information to an SNS topic.
+
+    We use this function to publish a message to the dispatch SNS topic, allowing
+    the raw Lambda to process this file and write it as an object to the
+    raw S3 bucket.
+
+    Args:
+        bucket (str): The input S3 bucket.
+        key (str): The S3 key of the export.
+        path (str): The file path within the export.
+        file_size (int): The size of the file in bytes.
+        sns_arn (str): The ARN of the dispatch SNS topic.
+
+    Returns:
+        None
+    """
+    sns_client = boto3.client("sns")
+    file_info = {
+        "Bucket": bucket,
+        "Key": key,
+        "Path": path,
+        "FileSize": file_size,
+    }
+    logger.info(f"Publishing {file_info} to {sns_arn}")
+    sns_client.publish(TopicArn=sns_arn, Message=json.dumps(file_info))
+
+
+def main(
+    event: dict,
+    s3_client: boto3.client,
+    input_bucket: str,
+    input_key_prefix: str,
+    raw_bucket: str,
+    raw_key_prefix: str,
+    dispatch_sns_arn: str,
+) -> None:
+    export_keys = list_s3_objects(
+        s3_client=s3_client,
+        bucket=input_bucket,
+        key_prefix=input_key_prefix,
+        key_format="input",
+    )
+    raw_keys = list_s3_objects(
+        s3_client=s3_client,
+        bucket=raw_bucket,
+        key_prefix=raw_key_prefix,
+        key_format="raw",
+    )
+    for export_key in sum(export_keys.values(), []):
+        # input bucket keys are formatted like `{namespace}/{cohort}/{export_basename}`
+        namespace, cohort = export_key.split("/")[:2]
+        file_list = list_files_in_archive(
```

Comment:
technically, should this ever be []? Should we be throwing some warning?

Reply:
CE shouldn't be sending us empty zip files, but agnostic of what's going on in the broader context of the pipeline, a function like `list_files_in_archive` should return an empty list if there are no files in the archive.

If we want a mechanism to alert us of suspiciously empty exports or other data validation issues with data in the input bucket, I don't think the raw-sync Lambda is the right context for that.

Reply:
I guess I'm also thinking about if somehow the logic in `list_files_in_archive` was inaccurate after updating it or what not, leading to a [] file. Should we flag that in `list_files_in_archive`?

Reply:
Are you thinking of what if there is a false negative? We return an empty list in two cases:

1. The EOCD marker is not found and the next `range_size` value we want to attempt is greater than 2x the object size, meaning we just downloaded the entire object and didn't find an EOCD marker. Hence, this is not a valid zip file and there's no file list to return.
2. We find an EOCD marker (this is a valid zip file), but either the zip file is empty, or it contains files which we don't want to process anyway according to our conditions:
```
"/" not in zip_info.filename  # necessary for pilot data only
and "Manifest" not in zip_info.filename
and zip_info.file_size > 0
```

We do `logger.error` in the case of (1), because we shouldn't pass non-zip files to `list_files_in_archive`. It could be useful to do something like `logger.warning` if `file_list` is still empty after (2), because an empty zip-file or a zip-file containing no useful data is pretty unusual, although it doesn't prevent `list_files_in_archive` from doing exactly what we expect of it. If there is a bug in the logic that the tests didn't catch, than a `logger.warning` could help us detect that.

Reply:
Sounds good to me!